File: python/privacy_kit/redaction.py
"""Minimización: detecta y anonimiza PII ANTES de enviarla a un LLM o a un tercero.

Es el punto de control más importante para sistemas con IA: evita que datos
personales (RUT, teléfono, tarjeta) viajen en claro a OpenAI/Anthropic.

Estrategia por defecto: pseudonimización reversible por sesión (tokens estables),
para que el LLM pueda razonar sobre "el cliente" sin ver el dato real, y luego
poder rehidratar si el flujo lo necesita (con control de acceso + auditoría).
"""
from __future__ import annotations
import re
import hashlib
from dataclasses import dataclass

from .config import PrivacyConfig


@dataclass
class RedactionResult:
    text: str                 # texto con PII reemplazada por tokens
    token_map: dict[str, str] # token -> valor original (NO enviar al LLM)
    found: dict[str, int]     # tipo_pii -> cantidad encontrada


class Redactor:
    def __init__(self, config: PrivacyConfig):
        self.patterns = {k: re.compile(v) for k, v in config.pii_patterns.items()}

    def _token(self, tipo: str, valor: str, salt: str) -> str:
        h = hashlib.sha256(f"{salt}:{valor}".encode()).hexdigest()[:8]
        return f"[{tipo.upper()}_{h}]"

    def redact(self, text: str, subject_id: str = "anon") -> RedactionResult:
        """Reemplaza PII por tokens estables. Devuelve texto seguro + mapa reversible.

        Enviar SOLO `result.text` al LLM. Guardar `result.token_map` en memoria de
        sesión protegida (nunca en logs ni al tercero).
        """
        if not text:
            return RedactionResult(text="", token_map={}, found={})
        token_map: dict[str, str] = {}
        found: dict[str, int] = {}
        out = text
        for tipo, rx in self.patterns.items():
            def _sub(m):
                val = m.group(0)
                tok = self._token(tipo, val, subject_id)
                token_map[tok] = val
                found[tipo] = found.get(tipo, 0) + 1
                return tok
            out = rx.sub(_sub, out)
        return RedactionResult(text=out, token_map=token_map, found=found)

    def rehydrate(self, text: str, token_map: dict[str, str]) -> str:
        """Reemplaza los tokens por sus valores reales (uso interno controlado)."""
        for tok, val in token_map.items():
            text = text.replace(tok, val)
        return text

File: node/src/redaction.js
'use strict';
// Minimización: detecta y anonimiza PII ANTES de enviarla a un LLM o a un tercero.
// Es el control más importante en sistemas con IA. Pseudonimización reversible por
// sesión: el LLM razona sobre "el cliente" sin ver el dato real; se rehidrata luego.
const crypto = require('crypto');

class Redactor {
  constructor(config) {
    this.patterns = {};
    for (const [tipo, rx] of Object.entries(config.piiPatterns || {})) {
      this.patterns[tipo] = new RegExp(rx, 'g');
    }
  }

  _token(tipo, valor, salt) {
    const h = crypto.createHash('sha256').update(`${salt}:${valor}`).digest('hex').slice(0, 8);
    return `[${tipo.toUpperCase()}_${h}]`;
  }

  /**
   * Reemplaza PII por tokens estables. Enviar SOLO `result.text` al LLM.
   * Guardar `result.tokenMap` en memoria de sesión protegida (nunca en logs ni al tercero).
   * @returns {{text:string, tokenMap:Object, found:Object}}
   */
  redact(text, subjectId = 'anon') {
    if (!text) return { text: '', tokenMap: {}, found: {} };
    const tokenMap = {};
    const found = {};
    let out = text;
    for (const [tipo, rx] of Object.entries(this.patterns)) {
      out = out.replace(rx, (m) => {
        const tok = this._token(tipo, m, subjectId);
        tokenMap[tok] = m;
        found[tipo] = (found[tipo] || 0) + 1;
        return tok;
      });
    }
    return { text: out, tokenMap, found };
  }

  /** Reemplaza los tokens por sus valores reales (uso interno controlado). */
  rehydrate(text, tokenMap) {
    for (const [tok, val] of Object.entries(tokenMap)) {
      text = text.split(tok).join(val);
    }
    return text;
  }
}

module.exports = { Redactor };
